### ISSUE TITLE
Address form not work properly in customer account area

### ISSUE THREAD

**Issue by metacreo** [PrestaShop/PrestaShop#34080]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

Tested on demo.prestashop.com
If you go to "account" and click update address (France) you need to enter states.
For some this bug exist only on Firefox or only on Chrome. On one my server with fresh PS install, bug seen in both browsers.
Reopen #33991.
On click Update on address in my-account area, form requesting enter state, even if country without states, Self-fixing only after manipulation with country change.

public function fillWith(array $params = []) called twice. Second call destroys country->id

### Expected behavior

No need enter states by default for France or other country with no states.

### Steps to reproduce
Clean browser cache before.
Go to demo.prestashop.com with 8.1.x version.
PS browser detection must be enabled.
PS must have enabled US country with states.
PS need to default US country (optional).
Ensure that your browser have default eu_US locale.
login as new customer and create address not US, all other without states.
Save address.
Click Update address.
Click Save.
 
https://github.com/PrestaShop/PrestaShop/assets/543458/a57f199d-05ec-4a6e-a859-14b1d1a78a2d

### PrestaShop version(s) where the bug happened

8.1.1 8.1.2

### PHP version(s) where the bug happened

8.2.10

**Comment by metacreo** [comment]
`public function fillWith(array $params = [])`
called twice, first call fill form with address, second call fill with blank default. that`s why states appear in form.
as temporary fix I write to my theme:

```
            {foreach from=$formFields item="field"}
              {if $field.name === 'id_state'}
                {if !Country::containsStates($formFields.id_country.value)}
                  {continue}
                {/if}
              {/if}
              {if $field.name === 'postcode'}
                {if !Country::getNeedZipCode($formFields.id_country.value)}
                  {$field.required = false}
                {/if}
              {/if}
              {block name='form_field'}
                {form_field field=$field}
              {/block}
            {/foreach}
```

**Comment by paulnoelcholot** [comment]
Hello @metacreo,

I tried to reproduce your issue but as you can see, I can't.

Did I miss something?

Thanks

[Identifiant.webm](https://github.com/PrestaShop/PrestaShop/assets/112863083/a653f139-3a1c-40ea-b3fb-de19169f9f13)

**Comment by metacreo** [comment]
@paulnoelcholot you not have any country that`s why this not issued on your site. go to demo.prestashop.com or try on shop who have more countries and default is country with states as default ps shop

**Comment by metacreo** [comment]
I don’t understand, you get paid to close real bugs? and don’t care that the bug is real.

**Comment by paulnoelcholot** [comment]
Please remain correct in your comments, thank you. There is a code of conduct and decorum to respect.

Thank you for your additional information, we will continue to investigate.

**Comment by metacreo** [comment]
you came here with one intention, to close the bug
initially, you are trying to repeat the bug in a modified store, knowing the problem in advance, you want to prove that bug does not exist.
Is there something wrong?

**Comment by matks** [comment]
@metacreo Please stop assuming bad intentions.

Every day plenty people report multiple bugs. Some are really bugs from the software. Some bugs come from a wrong configuration. Some bugs come from 3rd-party modules that are not part of the open source project. Some are not bugs, but simply wrong expectations (user expects software to do X but it does not do X, it does Y instead, and user assumes it means something is wrong).

Paulnoelcholot is doing a great job with other people of QA team of analyzing _every_ bug that is reported and trying to reproduce them in order to identify the real bugs from the software from all the other "bugs" (that are not). If he was not doing this, the project backlog would be drown with thousand of irrelevant bugs, making it completely useless. It's a meticulous work and it needs patience and knowledge.

**Comment by metacreo** [comment]
@matks 
I agree with you, the QA team does a lot of useful work, but unfortunately, not always.

A simple example is this bug. The man quickly realized the essence of the bug and tried to convince the public that there was no bug by disguising it. Why do you think he deliberately modified the settings to hide the bug? He deliberately removed all countries so that JavaScript would not drag the states into the form.
From here the only logical conclusion appears: he needs to close the bug and don’t give a damn about the PS project, he’s apparently just making money.

Previously, such minor bugs were resolved on the same day. Now it drags on for months. And all this thanks to your QA team. It's just a shame to see such a bug on such a platform. Most of the bugs (from 2.4k) have been fixed a long time ago or are not related to the engine and no one is working on them.

I have been developing on PS for more than 10 years, starting with version 1.5, and I know the objectmodel structure by heart. On my account there are unique top 1 projects in Google search results with 100k and more monthly rates.
This gives me the right to express my analytics and I see when QA works and when it doesn’t.

**Comment by paulnoelcholot** [comment]
Hello @metacreo 

Thank you very much for these great guesses about my intentions. I do this job by choice because having a quality application for users is my priority.
Your issue is incomplete and does not include a step to reproduce. I went to your previous issue to get the information needed to test. I failed to reproduce and my conclusion was "Did I miss something?" I have never expressed the fact of wanting to close the issue and even less hide anything for this purpose.
Namely, it is much easier and more beneficial in terms of time for me to validate an issue than to close an issue. Especially since if the bug is indeed present you will have to return to the subject again and again.

I just went to demo.prestashop.com, I didn't touch anything in the back office and I did the following test. I can't reproduce. Did I miss something? Can you please help me reproduce your bug so that I can ensure that the problem really comes from the application and that it can be handed over to the developers so that the fixes can benefit everyone?

Thank you in advance for your feedback.

[PrestaShop Live Demo.webm](https://github.com/PrestaShop/PrestaShop/assets/112863083/b5c15afd-9195-41ad-af55-d609b1588d1b)

**Comment by metacreo** [comment]
Hello @paulnoelcholot 
Thank you for trying help.
As I wrote above... it works in different ways. For example, I only have this bug on FireFox and it doesn’t exist on Chrome. For my colleague, on the contrary, the bug is observed only when using Chrome and not in FF.
Please try to check with different browsers.

**Comment by metacreo** [comment]
![photo_2023-10-03_13-40-29](https://github.com/PrestaShop/PrestaShop/assets/543458/3ef6f3e2-5b60-455e-9edd-92bdd99f9b92)
this screenshot from UK from my bro ))) (Chrome)
But in my Chrome this bug not seen.

**Comment by matks** [comment]
Please note that browsers differently validate forms. Chrome, Firefox, Edge, Safari... will sometimes differ about what is valid and what is not valid for a given form

**Comment by metacreo** [comment]
Yes, but template vars with states may not be assigned to form if country not contains states. Probably there is different works  JS.

**Comment by metacreo** [comment]
Found.
As mentioned above `public function fillWith(array $params = [])` called twice. Second call destroys country->id
classes/form/CustomerAddressForm.php
```
    public function fillWith(array $params = [])
    {
$trace = debug_backtrace();
dump("called by {$trace[1]['class']} :: {$trace[1]['function']}");
        // This form is tricky: fields may change depending on which country is being selected!
```
result:
`^ "called by CustomerAddressFormCore :: loadAddressById"`
`^ "called by AddressControllerCore :: postProcess"`
Fix:
AddressController.php Line 69
`//$this->address_form->fillWith(Tools::getAllValues());`
No bugs. Everything works.

**Comment by metacreo** [comment]
Need additional debugging. After 8.1.1 -> 8.1.2 my fix not work.

**Comment by metacreo** [comment]
top updated

Clean browser cache before.
Go to demo.prestashop.com with 8.1.x version.
PS browser detection must be enabled.
PS must have enabled US country with states.
Ensure that your browser have default eu_US locale.
login as new customer and create address not US, all other without states.
Save address.
Click Update address.
Click Save.

**Comment by Hlavtox** [comment]
Hi everyone, I managed to reproduce the issue. Maybe some steps to reproduce are extra, but I will narrow it down later.

### Setup
1. Install PrestaShop 8.1.2 in the language of your browser. I reproduce in Czech, but it will be French for example. This will be your home country.
2. Setup two countries. One will be your HOME country WITH STATES, the second country will be WITHOUT STATES, for example Germany.
3. Check that for example 2 States exist for your HOME country, check that if you go edit your HOME country, you have enabled "Contains states" and check that `State:name` is in the address format of that country.
4. Check that the second country does not have states.
5. Check that in localization settings, both "set country by browser" and "set language by browser" are enabled.

### Reproducing
1. Go to front office.
2. Register yourself as a customer.
3. Create a new address in Germany - the secound country.
4. Edit that country and see that the form contains State field with states for your home country. And you can't save the form.
5. The state field will disappear only after changing the country.

**Comment by infinityit2022** [comment]
> 5\. only after changing the co

Hi ,

**Comment by Maofree** [comment]
Hi @metacreo 
today I've found the same problem on prestashop 8.1.3 but the problem is not limited only to the selection of the state, in my case also of the dni field, but the problem is bigger, I believe that the fields used in the default nation are shown, in fact in my case if the dni field is removed from Italy then it is also removed from the other addresses already created, if I add it back then it also appears for those nations that do not have it among the chosen fields.
this is not a problem in the creation of addresses, but on those already created and which do not have the default country.
but here too I tested it by changing the default nation and the fields continue to refer to those of Italy which is the one chosen by the shopkeeper.

look at my topic

**Comment by Codencode** [comment]
The problem could be solved by executing the statement
```
$this->address_form->fillWith(Tools::getAllValues());
```
only if the form has been sent
So in theory it would be enough to add a check:
```
if (Tools::isSubmit('submitAddress')) {
         $this->address_form->fillWith(Tools::getAllValues());
}
```

The change affects the AddressControllerCore::postProcess() method

Now I create a Pull request like this.

**Comment by AureRita** [comment]
Hi @metacreo 

Thank you for your details, also you @Hlavtox. I can reproduce your issue with all that details as we can see : 

[recording(55).webm](https://github.com/PrestaShop/PrestaShop/assets/100699445/ae7d4d5d-d7eb-4c39-ad28-5c429038c592)

I'll add this to the backlog ( for other contributors ) 

Thank you @Codencode for your PR, I'll check it in few days / hour

The QA team will do everything it can to reproduce your issues and check that the PRs who correct them match your correction. Like you, we are human and we may have difficulty reproducing certain issues that are reported to us. 

In any case, thank you for helping us to improve Prestashop.

**Comment by metacreo** [comment]
Well done @AureRita Thank You.
Fix is here #34152
